statsmodels/distributions/copula.py

```diff
@@ -0,0 +1,227 @@
+def copula_bv_clayton(u, v, theta):
+    '''Clayton or Cook, Johnson bivariate copula
+    '''
+    if not theta > 0:
+        raise ValueError('theta needs to be strictly positive')
+    return np.power(np.power(u, -theta) + np.power(v, -theta) - 1, -theta)
+
+
+def copula_bv_frank(u, v, theta):
+    '''Cook, Johnson bivariate copula
+    '''
+    if not theta > 0:
+        raise ValueError('theta needs to be strictly positive')
+    cdfv = -np.log(1 + expm1(-theta*u) * expm1(-theta*v) / expm1(-theta))/theta
+    cdfv = np.minimum(cdfv, 1)  # necessary for example if theta=100
+    return cdfv
+
+
+def copula_bv_gauss(u, v, rho):
+    raise NotImplementedError
+
+
+def copula_bv_t(u, v, rho, df):
+    raise NotImplementedError
+
+
+# Archimedean Copulas through generator functions
+# ===============================================
+
+
+def copula_bv_archimedean(u, v, transform, args=()):
+    '''
+    '''
+    phi = transform.evaluate
+    phi_inv = transform.inverse
+    cdfv = phi_inv(phi(u, *args) + phi(v, *args), *args)
+    return cdfv
+
+
+def copula_mv_archimedean(u, transform, args=(), axis=-1):
+    '''generic multivariate Archimedean copula
+    '''
+    phi = transform.evaluate
+    phi_inv = transform.inverse
+    cdfv = phi_inv(phi(u, *args).sum(axis), *args)
+    return cdfv
+
+
+def copula_power_mv_archimedean(u, transform, alpha, beta, args=(), axis=-1):
+    '''generic multivariate Archimedean copula with additional power transforms
+
+    Nelson p.144, equ. 4.5.2
+    '''
+
+    def phi(u, alpha, beta, args=()):
+        return np.power(transform.evaluate(np.power(u, alpha), *args), beta)
+
+    def phi_inv(t, alpha, beta, args=()):
+        return np.power(transform.evaluate(np.power(t, 1. / beta), *args),
+                        1. / alpha)
+
+    cdfv = phi_inv(phi(u, *args).sum(axis), *args)
+    return cdfv
+
+
+class CopulaArchimedean(object):
+
+    def __init__(self, transform):
+        self.transform = transform
+
+    def cdf(self, u, args=(), axis=-1):
+        '''evaluate cdf of multivariate Archimedean copula
+        '''
+        phi = self.transform.evaluate
+        phi_inv = self.transform.inverse
+        cdfv = phi_inv(phi(u, *args).sum(axis), *args)
+        return cdfv
+
+    def pdf(self, u, args=(), axis=-1):
+        '''evaluate cdf of multivariate Archimedean copula
+        '''
+        u = np.asarray(u)
+        if u.shape[-1] > 2:
+            msg = "pdf is currently only available for bivariate copula"
+            raise ValueError(msg)
+        phi = self.transform.evaluate
+        phi_inv = self.transform.inverse
+        phi_d1 = self.transform.deriv
+        phi_d2 = self.transform.deriv2
+
+        cdfv = self.cdf(u, args=args, axis=axis)
+
+        pdfv = - np.product(phi_d1(u, *args), axis)
+        pdfv *= phi_d2(cdfv, *args)
+        pdfv /= phi_d1(cdfv, *args)**3
+
+        return pdfv
+
+
+# Extreme Value Copulas
+# =====================
+
+def copula_bv_ev(u, v, transform, args=()):
```

Comment:
Why not put these in their own module, so that you would have

distributions/copula/__init__.py
distributions/copula/bv.py
distributions/copula/mv.py
< and possibly others>

Reply:
Right now my plan was only to have two modules, not to build a full copula package, and I thought a longer namespace is not needed. If we plan to add nested and vine copulas, then this would get larger. But I will not be working on those.
(in an old private refactoring of sandbox.distributions, I had added subdirectories.)

bv and mv cannot be split nicely because it's mainly computational restrictions case by case, and I don't know yet how far we will get with mv.

Reply:
Checking my old code.
I have many subdirectories of sandbox.distributions, but most only contain one or two modules

Reply:
As a first step I would also restrict this to the bivariate case. Except for Gaussian, independent and t which can easily be n-dim. Otherwise it's a too big enterprise IMO